PARQUET-2117: Expose Row Index via ParquetReader and ParquetRecordReader

prakharjain09 opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/PARQUET-2117
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests:
Added TestParquetReader which covers rowIndex related tests for different kind of filters.
Also Extended all the ColumnIndexFiltering and BloomFiltering tests to validate the "row index" also. This adds unit test coverage for following scenarios for this feature: Parquet V1/V2 with encryption on/off with no-filter/simple-filter/column-index-filter/bloom-filter. 


### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

prakharjain09 commented on parquet-column/src/main/java/org/apache/parquet/column/page/PageReadStore.java:
```diff
@@ -43,6 +43,14 @@
+   *
+   * @return the row index offset of this row group.
+   */
+  default Optional<Long> getRowIndexOffset() {
```
@shangxinli I added this new method in the interface. This leads to build issues from japicmp-maven-plugin - METHOD_ADDED_TO_INTERFACE.

I provided a default value for this method and this seems to solve the issue. Please provide your feedback on this approach/any alternate approach?

prakharjain09 commented:
@shangxinli @gszadovszky Please review the changes when you get chance. Thanks!

shangxinli commented:
Can you squash the commits to make the review easier?

prakharjain09 commented:
> [](https://github.com/prakharjain09)Can you squash the commits to make the review easier?

done

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:
```diff
@@ -69,6 +71,8 @@
   private int currentBlock = -1;
   private ParquetFileReader reader;
+  private long currentRowIndex = -1L;
+  private PrimitiveIterator.OfLong rowIndexWithinFileIterator;
```
The name is so long

prakharjain09 replied:
renamed to `rowIdxInFileItr`.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:
```diff
@@ -265,4 +273,51 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
+   * Returns the ROW_INDEX of the current row.
+   */
+  public long getCurrentRowIndex() {
+    if (current == 0L) {
```
What is the reason not turning -1?

prakharjain09 replied:
current is an existing variable which tracks number of rows already processed. It is initialized to 0 at declaration time. So here we are trying to see if it is still 0, that means we haven't processed any row yet.

shangxinli replied:
I understand why we are checking 'current == 0L'. I was asking why you choose throw exception other than returning an invalid value. This is a public method. We should have it documented ether way you choose.

prakharjain09 replied:
There was no specific reason for choosing exception over -1.
I have updated it to return -1 and also updated all the public method docs to reflect the same.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:
```diff
@@ -265,4 +273,51 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
+   * Resets the row index iterator based on the current processed row group.
+   */
+  private void resetRowIndexIterator(PageReadStore pages) {
+    Optional<Long> rowIndexOffsetForCurrentRowGroup = pages.getRowIndexOffset();
```
The name is so long

prakharjain09 replied:
renamed to shorter name.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:
```diff
@@ -1400,34 +1400,67 @@ public ParquetMetadata readParquetMetadata(final InputStream from, MetadataFilte
+   * A container for [[FileMetaData]] and [[RowGroup]] to ROW_INDEX offset map.
+   */
+  private class FileMetaDataAndRowGroupOffsetInfo {
+    FileMetaData fileMetadata;
```
use final

prakharjain09 replied:
done.

shangxinli commented on parquet-hadoop/src/test/java/org/apache/parquet/filter2/recordlevel/PhoneBookWriter.java:
```diff
@@ -340,12 +342,21 @@ public static void write(ParquetWriter.Builder<Group, ?> builder, List<User> use
     return users;
   }
 
-  public static List<User> readUsers(ParquetReader.Builder<Group> builder) throws IOException {
```
Removing public method could cause incompatibility issue.

prakharjain09 replied:
fixed - not deleting it.

prakharjain09 commented on parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:
```diff
@@ -1400,34 +1400,67 @@ public ParquetMetadata readParquetMetadata(final InputStream from, MetadataFilte
     final byte[] encryptedFooterAAD = (encryptedFooter? AesCipher.createFooterAAD(fileDecryptor.getFileAAD()) : null);
 
-    FileMetaData fileMetaData = filter.accept(new MetadataFilterVisitor<FileMetaData, IOException>() {
+    FileMetaDataAndRowGroupOffsetInfo fileMetaDataAndRowGroupInfo = filter.accept(new MetadataFilterVisitor<FileMetaDataAndRowGroupOffsetInfo, IOException>() {
```
The `visit(OffsetMetadataFilter filter)` and `visit(RangeMetadataFilter filter)` returns the "filtered" FileMetadata and so few of the rowGroups might be missing in that. So doing it at the end will generate incorrect rowIndexOffset.

shangxinli replied:
Got it.

prakharjain09 commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:
```diff
@@ -248,15 +248,18 @@ public DictionaryPage readDictionaryPage() {
   private final RowRanges rowRanges;
 
-  public ColumnChunkPageReadStore(long rowCount) {
+  public ColumnChunkPageReadStore(long rowCount, long rowIndexOffset) {
```
Makes sense - retaining the older constructor.

prakharjain09 commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java:
```diff
@@ -265,6 +268,11 @@ public long getRowCount() {
 
+  @Override
+  public Optional<Long> getRowIndexOffset() {
+    return Optional.of(rowIndexOffset);
```
done.

prakharjain09 commented on parquet-hadoop/src/test/java/org/apache/parquet/filter2/recordlevel/PhoneBookWriter.java:
```diff
@@ -315,7 +317,7 @@ public static void write(ParquetWriter.Builder<Group, ?> builder, List<User> use
   }
 
-  private static ParquetReader<Group> createReader(Path file, Filter filter) throws IOException {
+  public static ParquetReader<Group> createReader(Path file, Filter filter) throws IOException {
```
This is being used by the new test file - https://github.com/apache/parquet-mr/pull/945/files#diff-276ac02899424a4245b8589f8ee6d444e3619f6be17834e4d5d7e81dfdaaee39R129

prakharjain09 commented:
@shangxinli Thanks fa lot or the review.

> [](https://github.com/prakharjain09)We need more test to cover old parquet data that doesn't have column index.

I couldn't find any existing tests or any existing parquet files in Resource directory which doesn't have column indexes. Could you please give some pointer to similar existing test or some way to create parquet file without column indexes (don't see any options to disable writing column indexes either)?
I have added tests for validating row indexes are correct with "column index filtering" disabled.

shangxinli commented:
I will have another look soon sometime this week.

prakharjain09 commented:
@shangxinli I have added test to cover old parquet file without column indexes. Please review the changes when you get chance.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java:
```diff
@@ -1476,6 +1509,13 @@ public ParquetMetadata fromParquetMetadata(FileMetaData parquetMetadata) throws
 
   public ParquetMetadata fromParquetMetadata(FileMetaData parquetMetadata,
       InternalFileDecryptor fileDecryptor, boolean encryptedFooter) throws IOException {
+    return fromParquetMetadata(parquetMetadata, fileDecryptor, encryptedFooter, generateRowGroupOffsets(parquetMetadata));
```
As you mentioned above, if parquetMetadata is a filtered one, then generateRowGroupOffsets() won't return accurate offsets, correct?

prakharjain09 replied:
Yes thats correct. Fixed this - now we are passing empty Map so that we don't populate incorrect rowIndexOffsets.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/metadata/BlockMetaData.java:
```diff
@@ -33,6 +33,7 @@
   private long totalByteSize;
   private String path;
   private int ordinal;
+  private long rowIndexOffset;
```
In the following toString(), it should be added too.

prakharjain09 replied:
done.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:
```diff
@@ -227,6 +232,9 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
 
         try {
           currentValue = recordReader.read();
+          if (rowIdxInFileItr != null) {
```
&& hasNext()?

prakharjain09 replied:
done.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:
```diff
@@ -265,4 +273,51 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
+      if (pages.getRowIndexes().isPresent()) {
+        rowIdxInRowGroupItr = pages.getRowIndexes().get();
+      } else {
+        // If `pages.getRowIndexes()` is empty, this means column indexing has not triggered.
```
The name of 'column index' was already used for Page Index in another feature. Can you use something else?

prakharjain09 replied:
removed this code comment.

prakharjain09 commented:
@shangxinli Thanks a lot or the review. I have addressed the review comments. Could you please look into the PR again?

Also could you share information about when are we planning to do code-freeze for next minor release? It will be great if we can release this change in next minor/patch release so that Apache Spark/other projects get to use this functionality sooner.

prakharjain09 commented:
> Could you please look into the PR again?
> Also could you share information about when are we planning to do code-freeze for next minor release? It will be great if we can release this change in next minor/patch release so that Apache Spark/other projects get to use this functionality sooner.


@shangxinli Gentle reminder. Please take a look when you get chance.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetReader.java:
```diff
@@ -140,6 +140,16 @@ public T read() throws IOException {
   }
 
+  /**
+   * Returns the row index of the last read row. If no row has been processed, returns -1.
```
Given this is public method, we need to take care of the Java doc decrations

shangxinli commented on parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestParquetReader.java:
```diff
@@ -46,10 +47,19 @@
   private static final Path FILE_V1 = createTempFile();
   private static final Path FILE_V2 = createTempFile();
-  private static final List<PhoneBookWriter.User> DATA = Collections.unmodifiableList(makeUsers(10000));
+  private static final Path STATIC_FILE_WITHOUT_COL_INDEXES = createPathFromCP("/test-file-with-no-column-indexes-1.parquet");
```
I am not sure if it is a good idea to check in a data file. Can you check if it is possible to stop generating offset index in the current version of Parquet?

prakharjain09 replied:
@shangxinli It looks like the [column-indexes](https://issues.apache.org/jira/browse/PARQUET-1201) are always written in the current version of parquet and are not configurable.
We are already testing the new row index support with and without the column index filtering being triggered (as part of TestColumnIndexFiltering). Also the new row index feature doesn't rely on column indexes in any way. So we can skip the backward compatibility testing and remove this parquet file from resources. What do you think about this?

shangxinli commented:
I just left some comments. Other than that, it looks good to me. Add @ggershinsky in case you have time to have a look.

ggershinsky commented:
hi guys, I'm OOO (vacation) this week. Can review it next week if helps, but feel free to go ahead without waiting for me.

prakharjain09 commented:
@shangxinli Thanks for taking another look. I have addressed all comments other [than one](https://github.com/apache/parquet-mr/pull/945#discussion_r820928524). Please advice on the same. Thanks!

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:
```diff
@@ -265,4 +275,46 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
+          return rowGroupRowIdxOffset.get() + rowIdxInRowGroupItr.next();
+        }
+      };
+    } else {
```
nit: could you start the method with checking this condition (!rowGroupRowIdxOffset.isPresent()), and then return? Will look cleaner.

prakharjain09 replied:
done.

ggershinsky commented:
thanks for this change. The PR looks good to me now, I'll add my approval after it passes the CI tests.

shangxinli commented:
@prakharjain09 After you fix the CI failures, we can merge.

prakharjain09 commented:
Thanks @ggershinsky for the review. I have addressed the comments and fixed the build issue.

prakharjain09 commented:
@shangxinli @ggershinsky Thanks a lot for reviewing this change.

This will unblock SPARK-37980 if this is released as part of upcoming parquet release. Do we need to cherry-pick this to any release branch for the same?

ggershinsky commented:
@prakharjain09 the upcoming parquet release will include the current master (plus a couple of WIP PRs, once they are merged), so this patch will be covered.

ggershinsky commented:
@prakharjain09 hopefully, we'll resolve the remaining issues at the community sync tomorrow, and start working on a cut.
